File: java/app/src/main/java/io/mikhail/components/StaticPage.java
package io.mikhail.components;

import com.pulumi.aws.s3.*;
import com.pulumi.aws.s3.inputs.*;
import com.pulumi.core.Output;
import com.pulumi.core.annotations.Export;
import com.pulumi.resources.ComponentResource;
import com.pulumi.resources.ComponentResourceOptions;
import com.pulumi.resources.CustomResourceOptions;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;

import java.util.List;
import java.util.Map;

public class StaticPage extends ComponentResource {
	@Export
	public final Output<String> endpoint;

	public StaticPage(String name, StaticPageArgs args, ComponentResourceOptions options) {
		super("static-page-component:index:StaticPage", name, options);

		// Create a bucket
		var bucket = new BucketV2(
				name + "-bucket",
				BucketV2Args.builder().build(),
				CustomResourceOptions.builder().parent(this).build());

		// Configure bucket website
		var bucketWebsite = new BucketWebsiteConfigurationV2(
				name + "-website",
				BucketWebsiteConfigurationV2Args.builder()
						.bucket(bucket.bucket())
						.indexDocument(BucketWebsiteConfigurationV2IndexDocumentArgs.builder()
								.suffix("index.html")
								.build())
						.build(),
				CustomResourceOptions.builder().parent(bucket).build());

		// Create bucket object for index document
		var indexObject = new BucketObject(
				name + "-index-object",
				BucketObjectArgs.builder()
						.bucket(bucket.bucket())
						.key("index.html")
						.content(args.indexContent())
						.contentType("text/html")
						.build(),
				CustomResourceOptions.builder().parent(bucket).build());

		// Create public access block
		var publicAccessBlock = new BucketPublicAccessBlock(
				name + "-public-access-block",
				BucketPublicAccessBlockArgs.builder()
						.bucket(bucket.bucket())
						.blockPublicAcls(false)
						.build(),
				CustomResourceOptions.builder().parent(bucket).build());

		// Create bucket policy
		var bucketPolicy = new BucketPolicy(
				name + "-bucket-policy",
				BucketPolicyArgs.builder()
						.bucket(bucket.bucket())
						.policy(bucket.bucket().applyValue(bucketName -> {
							var policy = Map.of(
									"Version", "2012-10-17",
									"Statement", List.of(
											Map.of(
													"Effect", "Allow",
													"Principal", "*",
													"Action", List.of("s3:GetObject"),
													"Resource", List.of(String.format("arn:aws:s3:::%s/*", bucketName))
											)
									)
							);
							try {
								return new ObjectMapper().writeValueAsString(policy);
							} catch (IOException e) {
								throw new RuntimeException("Failed to serialize policy to JSON", e);
							}
						}))
						.build(),
				CustomResourceOptions.builder()
						.parent(bucket)
						.dependsOn(publicAccessBlock)
						.build());

		this.endpoint = bucketWebsite.websiteEndpoint();

		this.registerOutputs(Map.of(
				"endpoint", this.endpoint
		));
	}
}

File: go/staticPage.go
package main

import (
	"encoding/json"
	"fmt"

	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/s3"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

type StaticPage struct {
	pulumi.ResourceState
	Endpoint pulumi.StringOutput `pulumi:"endpoint"`
}

type StaticPageArgs struct {
	IndexContent pulumi.StringInput `pulumi:"indexContent"`
}

func NewStaticPage(ctx *pulumi.Context, name string, args *StaticPageArgs, opts ...pulumi.ResourceOption) (*StaticPage, error) {
	comp := &StaticPage{}
	err := ctx.RegisterComponentResource("static-page-component:index:StaticPage", name, comp, opts...)
	if err != nil {
		return nil, err
	}

	// Create a bucket
	bucket, err := s3.NewBucketV2(ctx, fmt.Sprintf("%s-bucket", name), &s3.BucketV2Args{},
		pulumi.Parent(comp))
	if err != nil {
		return nil, err
	}

	// Configure bucket website
	bucketWebsite, err := s3.NewBucketWebsiteConfigurationV2(ctx, fmt.Sprintf("%s-website", name),
		&s3.BucketWebsiteConfigurationV2Args{
			Bucket: bucket.Bucket,
			IndexDocument: s3.BucketWebsiteConfigurationV2IndexDocumentArgs{
				Suffix: pulumi.String("index.html"),
			},
		},
		pulumi.Parent(bucket))
	if err != nil {
		return nil, err
	}

	// Create bucket object for index document
	_, err = s3.NewBucketObject(ctx, fmt.Sprintf("%s-index-object", name), &s3.BucketObjectArgs{
		Bucket:      bucket.Bucket,
		Key:         pulumi.String("index.html"),
		Content:     args.IndexContent,
		ContentType: pulumi.String("text/html"),
	}, pulumi.Parent(bucket))
	if err != nil {
		return nil, err
	}

	// Create public access block
	publicAccessBlock, err := s3.NewBucketPublicAccessBlock(ctx, fmt.Sprintf("%s-public-access-block", name),
		&s3.BucketPublicAccessBlockArgs{
			Bucket:          bucket.ID(),
			BlockPublicAcls: pulumi.Bool(false),
		}, pulumi.Parent(bucket))
	if err != nil {
		return nil, err
	}

	// Create bucket policy
	allowGetObjectPolicy := func(bucketName string) (string, error) {
		policy := map[string]interface{}{
			"Version": "2012-10-17",
			"Statement": []map[string]interface{}{
				{
					"Effect":    "Allow",
					"Principal": "*",
					"Action":    []string{"s3:GetObject"},
					"Resource":  []string{fmt.Sprintf("arn:aws:s3:::%s/*", bucketName)},
				},
			},
		}
		policyJSON, err := json.Marshal(policy)
		if err != nil {
			return "", err
		}
		return string(policyJSON), nil
	}

	_, err = s3.NewBucketPolicy(ctx, fmt.Sprintf("%s-bucket-policy", name), &s3.BucketPolicyArgs{
		Bucket: bucket.ID(),
		Policy: bucket.Bucket.ApplyT(func(bucketName string) (string, error) {
			return allowGetObjectPolicy(bucketName)
		}).(pulumi.StringOutput),
	}, pulumi.Parent(bucket), pulumi.DependsOn([]pulumi.Resource{publicAccessBlock}))
	if err != nil {
		return nil, err
	}

	// Set outputs
	comp.Endpoint = bucketWebsite.WebsiteEndpoint

	if err := ctx.RegisterResourceOutputs(comp, pulumi.Map{
		"endpoint": comp.Endpoint,
	}); err != nil {
		return nil, err
	}

	return comp, nil
}
